plugin/health_check.go

```diff
@@ -12,113 +12,82 @@ import (
+			activePlugins := job.env.Active()
+			for _, plugin := range activePlugins {
+				job.CheckPlugin(plugin.Manifest.Id)
 			}
+		case <-job.cancel:
+			return
 		}
-	}()
-}
-
-// checkPlugin determines the plugin's health status, then handles the error or success case.
-func (job *PluginHealthCheckJob) checkPlugin(id string) {
-	p, ok := job.env.registeredPlugins.Load(id)
-	if !ok {
-		return
-	}
-	rp := p.(registeredPlugin)
-
-	sup := rp.supervisor
-	if sup == nil {
-		return
-	}
-
-	pluginErr := sup.PerformHealthCheck()
-
-	if pluginErr != nil {
-		mlog.Error("Health check failed for plugin", mlog.String("id", id), mlog.Err(pluginErr))
-		job.handleHealthCheckFail(id, pluginErr)
 	}
 }
 
-// handleHealthCheckFail restarts or deactivates the plugin based on how many times it has failed in a configured amount of time.
-func (job *PluginHealthCheckJob) handleHealthCheckFail(id string, err error) {
-	rp, ok := job.env.registeredPlugins.Load(id)
-	if !ok {
+// CheckPlugin determines the plugin's health status, then handles the error or success case.
+// If the plugin passes the health check, do nothing.
+// If the plugin fails the health check, the function either restarts or deactivates the plugin, based on the quantity and frequency of its failures.
+func (job *PluginHealthCheckJob) CheckPlugin(id string) {
+	err := job.env.performHealthCheck(id)
+	if err == nil {
 		return
 	}
-	p := rp.(registeredPlugin)
 
-	// Append current failure before checking for deactivate vs restart action
-	p.failTimeStamps = append(p.failTimeStamps, time.Now())
-	p.lastError = err
-	job.env.registeredPlugins.Store(id, p)
+	mlog.Error("Health check failed for plugin", mlog.String("id", id), mlog.Err(err))
+	timestamps := job.getStoredTimestamps(id)
+	timestamps = append(timestamps, time.Now())
 
-	if shouldDeactivatePlugin(p) {
-		p.failTimeStamps = []time.Time{}
-		job.env.registeredPlugins.Store(id, p)
+	if shouldDeactivatePlugin(timestamps) {
+		// Order matters here, must deactivate first and then set plugin state
 		mlog.Debug("Deactivating plugin due to multiple crashes", mlog.String("id", id))
 		job.env.Deactivate(id)
-		job.env.SetPluginState(id, model.PluginStateFailedToStayRunning)
+
+		// Reset timestamp state for this plugin
+		job.failureTimestamps.Delete(id)
+		job.env.setPluginState(id, model.PluginStateFailedToStayRunning)
 	} else {
 		mlog.Debug("Restarting plugin due to failed health check", mlog.String("id", id))
 		if err := job.env.RestartPlugin(id); err != nil {
 			mlog.Error("Failed to restart plugin", mlog.String("id", id), mlog.Err(err))
 		}
+
+		// Store this failure so we can continue to monitor the plugin
+		job.failureTimestamps.Store(id, removeStaleTimestamps(timestamps))
 	}
 }
 
+// getStoredTimestamps returns the stored failure timestamps for a plugin.
+func (job *PluginHealthCheckJob) getStoredTimestamps(id string) []time.Time {
+	timestamps, ok := job.failureTimestamps.Load(id)
+	if !ok {
+		timestamps = []time.Time{}
```

Comment:
ahh sad that the other suggestion didn't work 😞 
```suggestion
		return nil
```